bmc/firmware.go

```diff
@@ -0,0 +1,156 @@
+package bmc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+// BMCVersionGetter retrieves the current BMC firmware version information
+type BMCVersionGetter interface {
+	GetBMCVersion(ctx context.Context) (version string, err error)
+}
+
+// BMCFirmwareUpdater upgrades the BMC firmware
+type BMCFirmwareUpdater interface {
+	FirmwareUpdateBMC(ctx context.Context, fileName string) (err error)
```

Comment:
might want to think about taking in an `io.Reader` or maybe `[]byte` so that a file could potentially come from anywhere instead of what appears here to be a location on disk. And then also return an `io.ReadCloser`. Just an thought

refs:
https://github.com/davecheney/reader/blob/master/reader.go
https://dave.cheney.net/2016/08/20/solid-go-design

Reply:
Thanks for pointing this out - I did give this a shot, turns out we need the filename, path when invoking the upload method as part of the multipart upload. As of now I'd like to take this up in another iteration.

https://github.com/bmc-toolbox/bmclib/pull/196/files#diff-ad657a5ff29b3a562ef3b7c45c08294e090a840dd10e2825da9f0e058f9cc59aR94